lutris/util/graphics/drivers.py

```diff
@@ -2,83 +2,139 @@
-                'date': ' '.join(nvrm_version[6:])
+            "nvrm": {
+                "vendor": nvrm_version[0],
+                "platform": nvrm_version[1],
+                "arch": nvrm_version[2],
+                "version": nvrm_version[5],
+                "date": " ".join(nvrm_version[6:]),
             }
         }
+    # If the /proc file failed, look it up with glxinfo.
+    glx_info = GlxInfo()
+    platform = read_process_output(["uname", "-s"])
+    arch = read_process_output(["uname", "-m"])
+    return {
+        "nvrm": {
+            "vendor": glx_info.opengl_vendor,  # type: ignore[attr-defined]
+            "platform": platform,
+            "arch": arch,
+            "version": glx_info.opengl_version.rsplit(maxsplit=1)[-1],  # type: ignore[attr-defined]
+        }
+    }
 
 
-def get_nvidia_gpu_ids():
+def get_nvidia_gpu_ids() -> List[str]:
     """Return the list of Nvidia GPUs"""
-    return os.listdir("/proc/driver/nvidia/gpus")
+    with contextlib.suppress(OSError):
+        return os.listdir("/proc/driver/nvidia/gpus")
+    values = read_process_output(
+        # 10de is NVIDIA's vendor ID, 0300 gets you video controllers.
+        ["lspci", "-D", "-n", "-d", "10de::0300"],
+    ).splitlines()
+    return [line.split(maxsplit=1)[0] for line in values]
 
 
-def get_nvidia_gpu_info(gpu_id):
+def get_nvidia_gpu_info(gpu_id: str) -> Dict[str, str]:
     """Return details about a GPU"""
-    with open("/proc/driver/nvidia/gpus/%s/information" % gpu_id, encoding='utf-8') as info_file:
-        content = info_file.readlines()
-    infos = {}
-    for line in content:
+    with contextlib.suppress(OSError):
+        with open(
+            "/proc/driver/nvidia/gpus/%s/information" % gpu_id, encoding="utf-8"
+        ) as info_file:
+            content = info_file.readlines()
+        infos = {}
+        for line in content:
+            key, value = line.split(":", 1)
+            infos[key] = value.strip()
+        return infos
+    lspci_data = read_process_output(["lspci", "-v", "-s", gpu_id])
+    model_info = re.search(r"NVIDIA Corporation \w+ \[(.+?)\]", lspci_data)
+    if model_info:
+        model = model_info.group(1)
+    irq_info = re.search("IRQ ([0-9]+)", lspci_data)
+    if irq_info:
+        irq = irq_info.group(1)
+
+    info = {
+        "Model": f"NVIDIA {model}",
+        "IRQ": irq,
+        "Bus Location": gpu_id,
+    }
+    for line in lspci_data.splitlines():
+        if ":" not in line:
+            continue
         key, value = line.split(":", 1)
-        infos[key] = value.strip()
-    return infos
+        info[key.strip()] = value.strip()
+    return info
 
 
-def is_nvidia():
+def is_nvidia() -> bool:
     """Return true if the Nvidia drivers are currently in use"""
-    return os.path.exists("/proc/driver/nvidia")
+    with contextlib.suppress(OSError):
+        return os.path.exists("/proc/driver/nvidia")
+    with contextlib.suppress(OSError):
```

Comment:
*Gah!* but maybe not so much *Gah!* here.

I still hate *contextlib.suppress* but this is trying a second path if the first fails, then falling back to GlxInfo() if they both fail.

It could be written straightforwardly as a pair of ``try: except OSError:``, but you maybe wind up nesting except blocks.

It would be appropriate to log *two* warnings here- one for each path. This assumes they are really separate things, but if they aren't why would we check both like this?

Reply:
Depending on the system's configuration, one may have access to to `/proc/modules` but not to `/proc/driver/nvidia`. For example, the current implementation of snap confinement blocks access to anything in `/proc/driver`, but not to `/proc/modules`, so an instance of Lutris confined within snap would use `/proc/modules` to detect it, but not `/proc/driver/nvidia`. However, some other configurations that still use NVIDIA's drivers may not have access to `/proc/modules` due to other security limitations. (For example, a multi-headed machine).

Realistically though, the `glxinfo` route here is the only one that can actually guarantee us that the user's session is using the NVIDIA drivers. Here are a few scenarios where the other two give the wrong answer:

1. A user has multiple graphics cards: An NVIDIA one purely for machine learning purposes, and an Intel one for desktop use and gaming. In this case, both the current implementation and the `/proc/modules` implementation would return `True` here, but the `glxinfo` way will return `False` since the NVIDIA drivers are loaded, but not being used to render anything.
2. A multi-headed machine with multiple graphics cards, including an NVIDIA one. If the user's session is rendered with a non-NVIDIA card.
3. (Not 100% sure about this since I don't have the hardware to test it, but I think...) An Optimus setup where the user has explicitly selected selected not to use the NVIDIA card for Lutris. (E.g. a developer who's trying to replicate an issue that occurs with Intel graphics.)

The code is unclear whether we're asking "is the nvidia kernel module loaded?" or "are we using nvidia drivers to render this session?" Both `/proc` based implementations answer the former, but I think what we actually care about for Lutris is the latter.

Without knowing more about the side effects I'm hesitant to commit to it, but if the latter is actually what we care about (which seems to be the case from the usages), I think just checking `glxinfo` would be the "most correct" way to do it.

Reply:
I'm not so sure. This code is used for checks at startup, but a laptop with hybrid graphics might start up in Intel mode, then trip into NVidia mode upon starting a game- long after Lutris startup. I think Lutris is asking if there's NVidia stuff around here, even if not in use right now.

Reply:
That makes sense. I'm reworking based on this info. Thanks!

Reply:
OK, so based on this I've made it:

1. Detect Nvidia drivers (2 different ways)
2. Fall back to `glxinfo` *but* log an error if it couldn't detect the module either way, since `glxinfo` has a potential to be wrong for what's being retrieved.

I also put in a note in a comment that `is_nvidia` doesn't detect nouveau drivers. Based on the previous function comment that appears to be intended, so I didn't change the behaviour.